tests/test_spatial_analysis/test_calculate_spatial_distance.py

```diff
@@ -0,0 +1,200 @@
+import unittest
+import anndata
+import numpy as np
+from spac.spatial_analysis import calculate_spatial_distance
+import pandas as pd
+import io
+from contextlib import redirect_stdout
+
+
+class TestCalculateSpatialDistance(unittest.TestCase):
+
+    def setUp(self):
+        # Create a sample AnnData object
+        self.adata = anndata.AnnData(X=np.random.rand(100, 10))
+        # Add spatial coordinates to adata.obsm
+        self.adata.obsm['spatial'] = np.random.rand(100, 2)
+        # Add annotations and image IDs to adata.obs
+        self.adata.obs['cell_type'] = np.random.choice(
+            ['type1', 'type2'], size=100
+        )
+        self.adata.obs['imageid'] = np.random.choice(
+            ['image1', 'image2'], size=100
+        )
+
+    def test_typical_case(self):
+        """Test the function with default parameters."""
+        calculate_spatial_distance(
+            adata=self.adata,
+            annotation='cell_type',
+            verbose=False
+        )
+        self.assertIn('spatial_distance', self.adata.uns)
+
+    def test_output_structure(self):
+        """Test that the output is stored as a pandas DataFrame."""
+        calculate_spatial_distance(
+            adata=self.adata,
+            annotation='cell_type',
+            verbose=False
+        )
+        result = self.adata.uns['spatial_distance']
+        self.assertIsInstance(result, pd.DataFrame)
+
+    def test_subset_processing(self):
+        """Test that the function correctly processes a subset of images."""
+        # Run the function with subset=['image1']
+        calculate_spatial_distance(
+            adata=self.adata,
+            annotation='cell_type',
+            subset=['image1'],
+            imageid='imageid',
+            verbose=False
+        )
+        result = self.adata.uns['spatial_distance']
+
+        # Get all cell IDs in 'image1' as strings
+        image1_cell_indices = self.adata.obs[
+            self.adata.obs['imageid'] == 'image1'
+        ].index.astype(str)
+
+        # Get the cell indices in the result
+        result_cell_indices = result.index.astype(str)
+
+        # Check that all cell IDs in result are from 'image1'
+        self.assertTrue(
+            set(result_cell_indices).issubset(set(image1_cell_indices))
+        )
+
+    def test_invalid_subset_type(self):
+        """Test that passing an invalid subset type raises a TypeError."""
+        # Pass an integer as subset, which should raise TypeError
+        with self.assertRaises(TypeError):
```

Comment:
@fangliu117 , please review the developer guide regarding how to test for error messages.

Reply:
Yes, I reviewed and understand to ensure accurate error messages are being generated, I should write unit tests to explicitly compare the expected error msg with the actual generated msg to ensure they match.